docs: add security docs (#715)
VictoriaMetrics/VictoriaLogs: docs/victorialogs/security.md

---
weight: 11
menu:
  docs:
    identifier: victorialogs-security
    parent: victorialogs
    weight: 11
title: Security
tags:
  - logs
---

VictoriaLogs implements multiple security layers to ensure data confidentiality and prevent unauthorized access.

> It supports [Basic Auth](https://docs.victoriametrics.com/victorialogs/security#Basic-Auth)

Author: Done

and [mTLS](https://docs.victoriametrics.com/victorialogs/security#mTLS) for user authentication and access control.
Traffic between clients and servers can be encrypted using [TLS/SSL](https://docs.victoriametrics.com/victorialogs/security#tlsssl).

Recommended practices:

* Deploy all VictoriaLogs components inside a private network, avoid direct Internet exposure.

> * Protect each component with Basic Auth + TLS or mTLS.

Author: Done, please review the changes

> * Use vmauth as an authorization proxy and load balancer between vlinsert and vlselect.

Author: Done, please review the changes

---
weight: 11
menu:
  docs:
    identifier: victorialogs-security
    parent: victorialogs
    weight: 11
title: Security
tags:
  - logs
---

VictoriaLogs implements multiple security layers to ensure data confidentiality and prevent unauthorized access.

> It supports [Basic Auth](https://docs.victoriametrics.com/victorialogs/security/#Basic-Auth)
> and [mTLS](https://docs.victoriametrics.com/victorialogs/security/#mTLS) for user authentication and access control.

Reviewer: ```suggestion
It supports [Basic Auth](https://docs.victoriametrics.com/victorialogs/security/#basic-auth)
and [mTLS](https://docs.victoriametrics.com/victorialogs/security/#mtls) for user authentication and access control.
```

Traffic between clients and servers can be encrypted using [TLS/SSL](https://docs.victoriametrics.com/victorialogs/security/#tlsssl).

Recommended practices:

* Deploy all VictoriaLogs components inside a private network, avoid direct Internet exposure.
* Use vmauth as an authorization proxy and load balancer in front of vlinsert and vlselect for authentication and authorization of data ingestion and querying requests.
  See [Security and Load Balancing](https://docs.victoriametrics.com/victorialogs/security-and-lb/) for configuration details.
  For TLS termination proxy setup, see [TLS termination proxy docs](https://docs.victoriametrics.com/victoriametrics/vmauth/#tls-termination-proxy).
* Protect components exposed to external clients with Basic Auth + TLS or mTLS.
* Disable read API on vlinsert with the `-select.disable` command-line flag.
* Disable write API on vlselect with the `-insert.disable` command-line flag.

## Basic Auth

It is recommended to enable Basic Auth when VictoriaLogs is deployed in untrusted or unsecured networks.

> For a full overview of security best practices, see the [security](https://docs.victoriametrics.com/victorialogs/security/#) section.

Reviewer: Wrong link?

```suggestion
For broader guidance, see [Security and Load Balancing](https://docs.victoriametrics.com/victorialogs/security-and-lb/).
```

Author: This link should redirect to the top of the page, specifically to the following list of best practices:

```
* Deploy all VictoriaLogs components inside a private network, avoid direct Internet exposure.
* Use vmauth as an authorization proxy and load balancer in front of vlinsert and vlselect for authentication and authorization of data ingestion and querying requests.
  See [Security and Load Balancing](https://docs.victoriametrics.com/victorialogs/security-and-lb/) for configuration details.
  For TLS termination proxy setup, see [TLS termination proxy docs](https://docs.victoriametrics.com/victoriametrics/vmauth/#tls-termination-proxy).
* Protect components exposed to external clients with Basic Auth + TLS or mTLS.
* Disable read API on vlinsert with the `-select.disable` command-line flag.
* Disable write API on vlselect with the `-insert.disable` command-line flag.
```

You don't need to enable Basic Auth in private networks - use it only in untrusted environments. That's what I wanted to say 😄

Reviewer: Hm weird, I changed it to localhost and clicked it, and it opened a new tab instead of scrolling to the top. If this doesn’t happen for you, please resolve it.

Basic Auth validates the `Authorization: Basic <base64(user:password)>` HTTP header.
If the username and password match the values configured in VictoriaLogs, the request is allowed.

Basic Auth encodes credentials using Base64 encoding, but does not encrypt the transmitted data.
Always use Basic Auth exclusively over [secure TLS/SSL connections](https://docs.victoriametrics.com/victorialogs/security/#tlsssl).

Production Security Recommendations:

- TLS encryption: never use Basic Auth over unencrypted HTTP connections.
- Consider mTLS: implement mutual TLS for certificate-based authentication instead of password-based auth.
- Regular credential rotation: change passwords periodically and use strong, unique passwords.

### Quick start

To quickly start VictoriaLogs with Basic Auth, run the following command:

```sh
./victoria-logs -httpAuth.username vlstorage -httpAuth.password secret
```

This will start the VictoriaLogs server with Basic authentication on the default port (9428).

> To test the configuration you can use `curl`:

Reviewer: ```suggestion
To test the configuration, you can use `curl`:
```

```sh
curl -u vlstorage:secret http://localhost:9428/metrics
```

For cluster mode, you need to configure both client authentication (users connecting to your service)
and internal authentication (communication between VictoriaLogs components):

```sh
./victoria-logs -httpListenAddr :9429 \
    -httpAuth.username vlinsert -httpAuth.password top-secret \
    -storageNode localhost:9428 \
    -storageNode.username vlstorage -storageNode.password secret
```

This configuration:
- Starts a VictoriaLogs server on port 9429 that functions as both vlinsert and vlselect

> - Accepts external requests only from user `vlselect` with password `top-secret`

Reviewer: ```suggestion
- Accepts external requests only from user `vlinsert` with password `top-secret`
```

- Connects to vlstorage at localhost:9428 using credentials `vlstorage:secret`

**Security Warning:** Passing passwords directly in command-line arguments is not recommended for production,
as they may appear in process listings (e.g., `ps aux`)
or debug endpoints (e.g., `http://victoria-logs:9428/debug/pprof/cmdline?debug=1`).

Use one of the safer methods below:

* [Providing password from a file](https://docs.victoriametrics.com/victorialogs/security/#providing-password-from-a-file).
* [Providing password via environment variables](https://docs.victoriametrics.com/victorialogs/security/#providing-password-via-environment-variables).
* [Providing password via HTTP/HTTPS](https://docs.victoriametrics.com/victorialogs/security/#providing-password-via-httphttps).

After configuring Basic Auth, update your clients accordingly, such as
vmauth (see [these docs](https://docs.victoriametrics.com/victorialogs/security-and-lb/))

> and VictoriaMetrics (see [these docs](https://docs.victoriametrics.com/victoriametrics/sd_configs/#http-api-client-options)).

Reviewer: ```suggestion
and VictoriaMetrics components (see [these docs](https://docs.victoriametrics.com/victoriametrics/sd_configs/#http-api-client-options)).
```

```

vlagent configuration example:

```sh
export VL_httpAuth_username=vlagent
export VL_httpAuth_password=top-top-secret
export VL_remoteWrite_basicAuth_username=vlinsert
export VL_remoteWrite_basicAuth_password=top-secret

./vlagent -envflag.enable -envflag.prefix VL_ \
    -httpListenAddr :9430 \
    -remoteWrite.url http://localhost:9429/internal/insert
```

To test that vlagent is properly configured with Basic Auth, you can send a test log entry:

```sh
curl -u vlagent:secret http://localhost:9430/insert/jsonline -H 'Content-Type: application/json' \
    -d '{"_msg":"Hello, VictoriaLogs!"}'
```

See also all supported command-line flags
for VictoriaLogs (see [these docs](https://docs.victoriametrics.com/victorialogs/#list-of-command-line-flags))
and vlagent (see [these docs](https://docs.victoriametrics.com/victorialogs/vlagent/#advanced-usage)).

### Providing password via HTTP/HTTPS

VictoriaLogs can retrieve passwords from HTTP/HTTPS endpoints for dynamic password management:

```sh
./victoria-logs -httpAuth.username foo \
    -httpAuth.password 'https://example.com/victoria-logs/password?example-arg=example-value'
```

> The value from the file is periodically reloaded, which allows changing the password without restarting the application.

Reviewer: ```suggestion
The value from the URL is periodically reloaded, which allows changing the password without restarting the application.
```

If using self-signed certificates, you have two options:
1. Install your CA certificate in the system's trusted certificate store.
1. Specify the CA certificate path manually using `-remoteWrite.tlsCAFile`.

For testing purposes only, you can disable certificate verification with `-remoteWrite.tlsInsecureSkipVerify`.

**Security Warning:** Disabling certificate verification eliminates
protection against man-in-the-middle attacks. Never use this in production.

### Automatic issuing of TLS certificates

All the [VictoriaLogs Enterprise](https://docs.victoriametrics.com/victoriametrics/enterprise/) components support automatic issuing of TLS certificates
for public HTTPS server running at `-httpListenAddr` via [Let's Encrypt service](https://letsencrypt.org/).
The following command-line flags must be set in order to enable automatic issuing of TLS certificates:

- `-httpListenAddr` must be set for listening TCP port `443`. For example, `-httpListenAddr=:443`.
  This port must be accessible by the [Let's Encrypt service](https://letsencrypt.org/).
- `-tls` must be set in order to accept HTTPS requests at `-httpListenAddr`.

>   Note that `-tlcCertFile` and `-tlsKeyFile` aren't needed when automatic TLS certificate issuing is enabled.

Reviewer: ```suggestion
  Note that `-tlsCertFile` and `-tlsKeyFile` aren't needed when automatic TLS certificate issuing is enabled.
```

- `-tlsAutocertHosts` must be set to comma-separated list of hosts, which can be reached via `-httpListenAddr`.
  TLS certificates are automatically issued for these hosts.
- `-tlsAutocertEmail` must be set to contact email for the issued TLS certificates.
- `-tlsAutocertCacheDir` may be set to the directory path for persisting the issued TLS certificates between VictoriaMetrics restarts.
  If this flag isn't set, then TLS certificates are re-issued on every restart.

Example of starting VictoriaLogs with automatic TLS certificate issuing:

```sh
./victoria-logs -httpListenAddr=:443 \
    -tls \
    -tlsAutocertHosts=victorialogs.example.com,logs.example.com \
    -tlsAutocertEmail=admin@example.com \
    -tlsAutocertCacheDir=/path/to/tls-cache \
    -licenseFile /path/to/license
```

#### The `no viable challenge type found` error

> Let's Encrypt validates domain ownership by performing HTTP-01 challenges to verify you control the specified domains.

Reviewer: for port 443?

```suggestion
Let's Encrypt validates domain ownership by performing TLS-ALPN-01 challenges to verify you control the specified domains.
```

The Let's Encrypt service must be able to reach your VictoriaLogs instance over the public internet on port 443 to complete this validation.

Common troubleshooting steps:
1. Ensure your server is reachable from the internet on port 443
1. Confirm that your domain names resolve to your server's public IP address

> 1. Use external tools to verify HTTP connectivity

Reviewer: port 443?

```suggestion
1. Use external tools to verify HTTPS connectivity
```

Example connectivity verification:
```sh
# Test domain resolution
nslookup victorialogs.example.com

# Test HTTP connectivity from external network
curl -I http://victorialogs.example.com/.well-known/acme-challenge/test
```

> This functionality can be evaluated for free according to [these docs](https://docs.victoriametrics.com/victoriametrics/enterprise/).

Author: This phrase was copied from another page and I can't find any information about free automatically issued TLS certificates. I suggest to remove this. Or I miss something?
